### PR TITLE
APEXCORE-532: Fix issue where new operators added to dag starts from initial checkpoint

### PR DESCRIPTION
**Pull request by tushargosavi**
In case new operators are added dynamically, they should start from windowId of upstream operator. The issue was that new operators gets added using addLogicalOpeartor which sets recoveryWindowId to INITIAL_CHECKPOINT, which prevents getActivationCheckpoint to return correct window id of the operator during deploy.

@PramodSSImmaneni  please review.

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 1]
```diff
@@ -1264,12 +1264,12 @@ public Operator loadOperator(PTOperator oper)
    */
   private Checkpoint getActivationCheckpoint(PTOperator oper)
   {
-    if (oper.recoveryCheckpoint == null && oper.checkpoints.isEmpty()) {
+    if ((oper.recoveryCheckpoint == null || oper.recoveryCheckpoint == Checkpoint.INITIAL_CHECKPOINT) && oper.checkpoints.isEmpty()) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: It was supposed to be noop for an already initialized operator. Why is recoveryCheckpoint assigned in your case? There is the assumption in initPartitioning that does not seem correct.

**Comment by pramodin** [comment]
can you add a unit test

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 1 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: When new operator is added to DAG its get added through addLogicalOperator, which
calls initParitioning with INITIAL_CHECKPOINT , which calls addPTOperator which
sets recoveryCheckpoint to INITIAL_CHECKPOINT. It happens during new DAG start also,
all operators recoveryCheckpoint is set to INITIAL_CHECKPOINT.

only in case of newly added unifiers, recoveryCheckpoint is set to null, hence getActivationCheckpoint returns correct windowId, and initCheckpoint is called again with new windowId.

**Comment by tushargosavi** [comment]
working on a test case.

**Comment by tushargosavi** [comment]
@PramodSSImmaneni I have added a test case

File: engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java [hunk 2]
```diff
@@ -408,4 +409,47 @@ public void testExecutionManagerWithAsyncStorageAgent() throws Exception
     testExecutionManager(new AsyncFSStorageAgent(testMeta.getPath(), null));
   }
 
+  @Test
+  public void testNewOpearatorRecoveryWindowIds()
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java]: spelling

File: engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java [hunk 3]
```diff
@@ -408,4 +409,47 @@ public void testExecutionManagerWithAsyncStorageAgent() throws Exception
...
+    pm.addStream("s2", o2.outport1, o3.inport1);
+
+    Assert.assertEquals("undeploy " + ctx.undeploy, 0, ctx.undeploy.size());
+    Assert.assertEquals("deploy " + ctx.deploy, 0, ctx.deploy.size());
+
+    pm.applyChanges(ctx);
+
+    Assert.assertEquals("containers post change", 3, plan.getContainers().size());
+
+    Assert.assertEquals("undeploy " + ctx.undeploy, 1, ctx.undeploy.size());
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java]: How are these related to this test?

File: engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java [hunk 4]
```diff
@@ -408,4 +409,47 @@ public void testExecutionManagerWithAsyncStorageAgent() throws Exception
...
+    Assert.assertEquals("deploy " + ctx.deploy, 3, ctx.deploy.size());
+
+    LogicalPlan.OperatorMeta o2Meta = plan.getLogicalPlan().getMeta(o2);
+    List<PTOperator> o2Partitions = plan.getOperators(o2Meta);
+    Assert.assertEquals("unifier activation checkpoint " + o2Meta, 10, o2Partitions.get(0).getRecoveryCheckpoint().windowId);
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java]: unifier?

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 1 above)
Comment (tweise) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: I believe this check should be removed and instead null passed to addPTOperator. Please verify.

```
+++ b/engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java
@@ -781,7 +781,7 @@ public class PhysicalPlan implements Serializable
     // create operator instance per partition
     Map<Integer, Partition<Operator>> operatorIdToPartition = Maps.newHashMapWithExpectedSize(partitions.size());
     for (Partition<Operator> partition : partitions) {
-      PTOperator p = addPTOperator(m, partition, Checkpoint.INITIAL_CHECKPOINT);
+      PTOperator p = addPTOperator(m, partition, null);
       operatorIdToPartition.put(p.getId(), partition);
     }

@@ -1264,7 +1264,7 @@ public class PhysicalPlan implements Serializable
    */
   private Checkpoint getActivationCheckpoint(PTOperator oper)
   {
-    if ((oper.recoveryCheckpoint == null || oper.recoveryCheckpoint == Checkpoint.INITIAL_CHECKPOINT) && oper.checkpoints.isEmpty()) {
+    if (oper.recoveryCheckpoint == null && oper.checkpoints.isEmpty()) {
       Checkpoint activationCheckpoint = Checkpoint.INITIAL_CHECKPOINT;
       for (PTInput input : oper.inputs) {
         PTOperator sourceOper = input.source.source;
```

**Review by tweise — CHANGES_REQUESTED** [review]

**Comment by tushargosavi** [comment]
Verifying the suggested change

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 5]
```diff
@@ -1268,10 +1268,11 @@ private Checkpoint getActivationCheckpoint(PTOperator oper)
       Checkpoint activationCheckpoint = Checkpoint.INITIAL_CHECKPOINT;
       for (PTInput input : oper.inputs) {
         PTOperator sourceOper = input.source.source;
+        Checkpoint checkPoint = sourceOper.recoveryCheckpoint;
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: minor: can you change the variable name to checkpoint

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: renamed variable name to checkpoint

File: engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java (hunk 3 above)
Comment (tushargosavi) [reply engine/src/test/java/com/datatorrent/stram/LogicalPlanModificationTest.java]: removed these

**Comment by tushargosavi** [comment]
@tweise I have updated the pull request
